common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/challengehandlers/PKeyAuthChallengeHandler.java

```diff
@@ -22,21 +22,259 @@
+     *                           &CertAuthorities=<distinguished names of CAs>&Version=1.0
+     *                           &SubmitUrl=<URL to submit response>&Context=<server state that
+     *                           client must convey back>
+     * @param view
+     * @param completionCallback
+     * @throws ClientException
+     */
+    public PKeyAuthChallengeHandler(@NonNull final String redirectUri,
+                                    @NonNull final WebView view,
+                                    @NonNull final AuthorizationRequest request,
+                                    @NonNull IChallengeCompletionCallback completionCallback) {
+        mWebView = view;
+        mRedirectUri = redirectUri;
+        mRequest = request;
+        mChallengeCallback = completionCallback;
+    }
+
+    public void process() {
+        mWebView.stopLoading();
+        mChallengeCallback.setPKeyAuthStatus(true);
+        new Thread(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    final PkeyAuthChallengeResponse challengeResponse = getChallengeResponseFromUri();
+                    final Map<String, String> headers = new HashMap<>();
+                    headers.put(AuthenticationConstants.Broker.CHALLENGE_RESPONSE_HEADER,
+                            challengeResponse.getAuthorizationHeaderValue());
+                    mWebView.post(new Runnable() {
+                        @Override
+                        public void run() {
+                            String loadUrl = challengeResponse.getSubmitUrl();
+                            Logger.verbose(TAG, "Respond to pkeyAuth challenge");
+                            Logger.verbosePII(TAG, "Challenge submit url:" + challengeResponse.getSubmitUrl());
+
+                            mWebView.loadUrl(loadUrl, headers);
+                        }
+                    });
+                } catch (final ClientException e) {
+                    // It should return error code and finish the
+                    // activity, so that onActivityResult implementation
+                    // returns errors to callback.
+                    Intent resultIntent = new Intent();
+                    resultIntent.putExtra(AuthenticationConstants.Browser.RESPONSE_AUTHENTICATION_EXCEPTION, e);
+                    resultIntent.putExtra(AuthenticationConstants.Browser.RESPONSE_REQUEST_INFO, mRequest);
+                    mChallengeCallback.sendResponse(AuthenticationConstants.UIResponse.BROWSER_CODE_AUTHENTICATION_EXCEPTION, resultIntent);
+                }
+            }
+        }).start();
+    }
+
+    /**
+     * This parses the redirectURI for challenge components and produces
+     * response object.
+     *
+     * @return Return Device challenge response
+     */
+    private PkeyAuthChallengeResponse getChallengeResponseFromUri() throws ClientException {

```

Comment:
This is basically a factory method but because it's `private` we cannot test it -- how do you know this works??